OpenSim/Simulation/Model/Geometry.cpp

```diff
@@ -171,50 +171,63 @@ void FrameGeometry::implementCreateDecorativeGeometry(SimTK::Array_<SimTK::Decor
 
 void Mesh::implementCreateDecorativeGeometry(SimTK::Array_<SimTK::DecorativeGeometry>& decoGeoms) const
 {
-    const std::string& file = get_mesh_file();
-    bool isAbsolutePath; string directory, fileName, extension;
-    SimTK::Pathname::deconstructPathname(file,
-        isAbsolutePath, directory, fileName, extension);
-    const string lowerExtension = SimTK::String::toLower(extension);
-    if (lowerExtension != ".vtp" && lowerExtension != ".obj" && lowerExtension != ".stl") {
-        std::clog << "ModelVisualizer ignoring '" << file
-            << "'; only .vtp .stl and .obj files currently supported.\n";
-        return;
-    }
-
-    // File is a .vtp or .obj. See if we can find it.
-    Array_<string> attempts;
-    const Model& model = getFrame().getModel();
-    bool foundIt = ModelVisualizer::findGeometryFile(model, file, isAbsolutePath, attempts);
-
-    if (!foundIt) {
-        std::clog << "ModelVisualizer couldn't find file '" << file
-            << "'; tried\n";
-        for (unsigned i = 0; i < attempts.size(); ++i)
-            std::clog << "  " << attempts[i] << "\n";
-        if (!isAbsolutePath &&
-            !Pathname::environmentVariableExists("OPENSIM_HOME"))
-            std::clog << "Set environment variable OPENSIM_HOME "
-            << "to search $OPENSIM_HOME/Geometry.\n";
-        return;
-    }
-
-    SimTK::PolygonalMesh pmesh;
-    try {
+    // Current interface to Visualizer calls this method on every frame.
+    // On first time through, load file and create DecorativeMeshFile else
+    // try to reuse it.
+    if (!triedToLoad) {
+        triedToLoad = true;
+        const std::string& file = get_mesh_file();
+        bool isAbsolutePath; string directory, fileName, extension;
+        SimTK::Pathname::deconstructPathname(file,
+            isAbsolutePath, directory, fileName, extension);
+        const string lowerExtension = SimTK::String::toLower(extension);
+        if (lowerExtension != ".vtp" && lowerExtension != ".obj" && lowerExtension != ".stl") {
+            std::clog << "ModelVisualizer ignoring '" << file
+                << "'; only .vtp .stl and .obj files currently supported.\n";
+            return;
+        }
```

Comment:
Why are you trying to load geometry at drawing time? This seems like a good case for extending finalize from properties to handle the geometry loading prior to simulation and visualization time. It would push these checks earlier where they belong IMHO.